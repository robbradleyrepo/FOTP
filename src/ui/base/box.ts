import styled from "styled-components";

import helpers from "../styles/helpers";

const Box = styled.div`
  ${helpers};
`;

export default Box;
