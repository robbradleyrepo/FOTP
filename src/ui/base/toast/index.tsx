import { useLocale } from "@sss/i18n";
import { AnimatePresence, motion } from "framer-motion";
import React, {
  createContext,
  FC,
  MouseEventHandler,
  ReactNode,
  useContext,
  useEffect,
  useRef,
  useState,
} from "react";

import {
  ComponentStyleProps,
  px,
  py,
  s,
  size,
  useTheme,
} from "@/common/ui/utils";

import cross from "../../icons/cross";
import Icon from "../icon/";

const enUsResource = {
  controls: {
    close: "Close",
  },
};

export enum ToastType {
  ERROR = "ERROR",
  INFO = "INFO",
  WARNING = "WARNING",
}

export const TOAST_TIMEOUT = {
  [ToastType.ERROR]: 5000,
  [ToastType.INFO]: 8000,
  [ToastType.WARNING]: 8000,
};

export interface ToastData {
  children: ReactNode;
  type?: ToastType;
}

interface ToastDataExtended extends ToastData {
  id: string;
  type: ToastType;
}

interface ToastContextProps {
  current?: ToastDataExtended;
  dismiss: MouseEventHandler;
  push: (data: ToastData) => void;
}

const ToastContext = createContext<ToastContextProps | null>(null);

export const ToastController: FC = ({ children }) => {
  const { i18n } = useLocale();
  const toastArrayRef = useRef<ToastDataExtended[]>([]);
  const [current, setCurrent] = useState<ToastDataExtended | undefined>(
    toastArrayRef.current[0]
  );

  i18n.addResourceBundle("en-US", "toast", enUsResource);

  const update = () => setCurrent(toastArrayRef.current[0]);

  const dismiss = () => {
    toastArrayRef.current = toastArrayRef.current.slice(1);
    update();
  };

  const push = ({ children, type = ToastType.INFO }: ToastData) => {
    toastArrayRef.current = [
      ...toastArrayRef.current,
      {
        children,
        id: Math.random().toString(16).substr(2, 8), // Generate a random ID so we can distinguish between toasts
        type,
      },
    ];
    update();
  };

  useEffect(() => {
    if (!current) {
      return;
    }

    const timerId = setTimeout(dismiss, TOAST_TIMEOUT[current.type]);

    return () => clearTimeout(timerId);
  }, [current]);

  return (
    <ToastContext.Provider value={{ current, dismiss, push }}>
      {children}
    </ToastContext.Provider>
  );
};

export const useToastController = () => {
  const context = useContext(ToastContext);

  if (!context) {
    throw new Error(
      "`useToastController` must be used inside a `ToastController`"
    );
  }

  return context;
};

interface ToastProps extends ToastData {
  onClose: MouseEventHandler;
}

const Toast: FC<ToastProps> = ({
  children,
  onClose,
  type = ToastType.INFO,
}) => {
  const { t } = useLocale();
  const theme = useTheme();

  const backgroundColor = {
    [ToastType.ERROR]: theme.color.state.error,
    [ToastType.INFO]: theme.color.background.dark,
    [ToastType.WARNING]: theme.color.state.warning,
  }[type];

  return (
    <div
      css={s((t) => ({
        backgroundColor: backgroundColor,
        borderRadius: t.radius.xs,
        boxShadow:
          "0px 3px 4px rgba(0, 0, 0, 0.15), 0px 0px 10px rgba(0, 0, 0, 0.25)",
        color: t.color.text.light.base,
        paddingLeft: t.spacing.sm,
        paddingRight: t.spacing.lg,
        ...py(12),
        pointerEvents: "auto",
        position: "relative",
      }))}
      role="alert"
    >
      {children}
      <button
        css={s({
          padding: 4,
          position: "absolute",
          right: 14,
          ...size(20),
          top: 14,
        })}
        onClick={onClose}
      >
        <Icon
          _css={s({
            verticalAlign: "top",
          })}
          path={cross}
          title={t("toast:controls.close")}
        />
      </button>
    </div>
  );
};

export const ToastRack: FC<ComponentStyleProps> = ({ _css }) => {
  const { current, dismiss } = useToastController();

  return (
    <div
      css={s(
        {
          maxWidth: 400,
          pointerEvents: "none",
          position: "relative",
          width: "100%",
        },
        _css ?? {}
      )}
    >
      <AnimatePresence>
        {current && (
          <motion.div
            css={s((t) => ({
              paddingTop: [t.spacing.xs, null, t.spacing.sm],
              ...px([t.spacing.sm, null, t.spacing.lg]),
              position: "absolute",
              width: "100%",
            }))}
            key={current.id}
            animate={{ opacity: 1, transition: { duration: 0.5 }, x: 0, y: 0 }}
            exit={{ opacity: 0, transition: { duration: 0.3 }, y: "-100%" }}
            initial={{ opacity: 0, x: "100%" }}
            transition={{ ease: "anticipate" }}
          >
            <Toast {...current} onClose={dismiss} />
          </motion.div>
        )}
      </AnimatePresence>
    </div>
  );
};
