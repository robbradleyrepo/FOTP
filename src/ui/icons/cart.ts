export default "M19.7 4.7c-.1-.1-.1-.2-.1-.2l-3-4c-.1-.2-.4-.3-.6-.3H4c-.2 0-.5.1-.6.3l-3 4s0 .1-.1.1v.1c0 .1-.1.2-.1.3v14a2.732 2.732 0 002.7 2.7h14a2.732 2.732 0 002.7-2.7V5c.2-.1.1-.2.1-.3zM4.4 1.8h11.2l1.9 2.5h-15l1.9-2.5zM18.3 19c-.1.3-.2.7-.5.9-.3.2-.6.4-.9.4H3c-.3 0-.7-.1-.9-.4-.2-.3-.3-.6-.3-.9V5.8h16.5V19z M10 13.8c1.2 0 2.5-.5 3.4-1.4.9-.9 1.4-2.1 1.4-3.4 0-.4-.3-.8-.8-.8s-.8.4-.8.8c0 .9-.3 1.7-1 2.3-1.2 1.2-3.4 1.2-4.6 0-.6-.6-1-1.4-1-2.3 0-.4-.3-.8-.8-.8s-.6.4-.6.8c0 1.3.5 2.5 1.4 3.4.9.9 2.1 1.4 3.4 1.4z";
