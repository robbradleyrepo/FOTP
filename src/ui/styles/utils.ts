import { color } from "./variables";

interface DataUriFromPathParams {
  fill?: string;
  path: string;
  viewBox?: string;
  [k: string]: number | string | undefined;
}

const dataUriFromPath = ({
  fill = color.text.dark.base,
  path,
  viewBox = "0 0 100 100",
  ...rest
}: DataUriFromPathParams) => {
  const attributes = Object.entries(rest)
    .map(([key, value]) => ` ${key}="${value}"`)
    .join("");

  return `data:image/svg+xml;charset=utf8,${encodeURIComponent(`<svg viewBox="${viewBox}" xmlns="http://www.w3.org/2000/svg"${attributes}><path d="${path}" fill="${fill}" /></svg>
`)}`;
};

export { dataUriFromPath };
