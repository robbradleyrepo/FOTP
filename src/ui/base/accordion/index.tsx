import { motion } from "framer-motion";
import React, {
  ComponentType,
  FC,
  MouseEventHandler,
  ReactNode,
  useState,
} from "react";

import { ComponentStyleProps, mx, py, s, size } from "@/common/ui/utils";

import chevronDown from "../../icons/chevronDown";
import Icon from "../icon";
import StyledComponentsHelper from "../styled-components-helper";
import { bodyText } from "../typography";

interface CommonAccordionProps extends ComponentStyleProps {
  id: string;
  label: ReactNode;
  labelAs: keyof JSX.IntrinsicElements | ComponentType;
}
export interface AccordionUIProps extends CommonAccordionProps {
  isOpen: boolean;
  onClick: MouseEventHandler;
}

// Based on https://www.w3.org/TR/wai-aria-practices-1.1/#accordion
export const AccordionUI: FC<AccordionUIProps> = ({
  _css = {},
  children,
  id,
  isOpen,
  label,
  labelAs,
  onClick,
}) => {
  /*
   *  Although we don't need to access these IDs outside of this component
   *  it's really difficult to automatically generate unique IDs that are
   *  consistent between server and client - see
   *  https://github.com/facebook/react/issues/15435 and
   *  https://github.com/facebook/react/pull/17322
   */
  const contentId = `${id}-content`;
  const labelId = id;

  return (
    <div
      css={s(
        bodyText,
        {
          borderBottomStyle: "solid",
          borderColor: "rgba(47, 78, 37, 0.15)",
          borderWidth: 1,
        },
        _css
      )}
    >
      <StyledComponentsHelper
        as={
          // We need to cast to any to avoid an `Type instantiation is
          // excessively deep and possibly infinite` error - see
          // https://github.com/microsoft/TypeScript/issues/34933
          labelAs as any // eslint-disable-line @typescript-eslint/no-explicit-any
        }
        css={s((t) => py(t.spacing.xs))}
      >
        <button
          css={s((t) => ({
            alignItems: "center",
            display: "flex",
            justifyContent: "space-between",
            ...py(t.spacing.sm),
            width: "100%",
          }))}
          aria-controls={contentId}
          aria-expanded={isOpen}
          id={labelId}
          onClick={onClick}
        >
          {label}
          <Icon
            _css={s((t) => ({
              flexShrink: 0,
              ...mx(t.spacing.sm),
              ...size(14),
              transform: `rotate(${isOpen ? "-180deg" : "0"})`, // Use 2D transform until Intercom remove negative z-index - see https://community.intercom.com/t/z-index-issue-with-ios-13/1432
              transition: "transform 500ms",
            }))}
            path={chevronDown}
          />
        </button>
      </StyledComponentsHelper>
      <motion.div
        css={s({ overflow: "hidden" })}
        animate={isOpen ? "open" : "closed"}
        aria-labelledby={labelId}
        hidden={!isOpen}
        id={contentId}
        role="region"
        transition={{
          damping: 100,
          mass: 0.5,
          stiffness: 1000,
          type: "spring",
        }}
        variants={{
          closed: { height: 0, transitionEnd: { display: "none" } },
          open: { display: "block", height: "auto" },
        }}
      >
        {children}
      </motion.div>
    </div>
  );
};

export interface AccordionProps extends CommonAccordionProps {
  initiallyOpen?: boolean;
  onClick?: MouseEventHandler;
}

const Accordion: FC<AccordionProps> = ({
  initiallyOpen = false,
  onClick,
  ...rest
}) => {
  const [isOpen, setIsOpen] = useState(initiallyOpen);

  const handleClick: MouseEventHandler = async (...args) => {
    setIsOpen(!isOpen);

    if (onClick) {
      return onClick(...args);
    }
  };

  return <AccordionUI {...rest} isOpen={isOpen} onClick={handleClick} />;
};

export default Accordion;
