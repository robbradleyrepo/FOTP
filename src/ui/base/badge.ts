import { px, py, s } from "@/common/ui/utils";

import { headingEcho } from "./typography";

export const badge = s(headingEcho, (t) => ({
  backgroundColor: t.color.tint.lightGreen,
  borderRadius: t.radius.xs,
  color: t.color.text.dark.base,
  fontSize: [12, null, 14],
  ...px(t.spacing.xs),
  ...py(1),
  whiteSpace: "nowrap",
}));
