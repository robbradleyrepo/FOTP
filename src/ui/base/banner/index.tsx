import { px, py, s } from "@/common/ui/utils";

import { height } from "../../styles/variables";
import { bodyText } from "../typography";

const banner = s(bodyText, (t) => ({
  alignItems: "center",
  backgroundColor: t.color.background.dark,
  color: t.color.text.light.base,
  display: "flex",
  fontSize: [14, 16, 18],
  height: [height.banner.mobile, null, height.banner.desktop],
  justifyContent: "center",
  left: 0,
  lineHeight: ["18px", null, null, "22px"],
  position: "fixed",
  ...px(t.spacing.md),
  ...py(t.spacing.md),
  right: 0,
  textAlign: "center",
  top: 0,
  width: "100vw", // Use `vw` instead of percentage to match the body width with `scroll: overlay`
  zIndex: 999,
}));

export default banner;
