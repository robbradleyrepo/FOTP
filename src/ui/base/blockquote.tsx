import React, { FC, ReactNode } from "react";

import { belt, ComponentStyleProps, px, s } from "@/common/ui/utils";

import { bodyText, headingAlpha, headingCharlie } from "./typography";

interface BlockquoteProps extends ComponentStyleProps {
  attribution?: ReactNode;
  ornament?: ReactNode;
}

const Blockquote: FC<BlockquoteProps> = ({
  _css = {},
  attribution,
  children,
  ornament,
}) => (
  <figure
    css={s(
      belt,
      (t) => ({
        borderColor: t.color.background.feature3,
        borderStyle: "solid",
        borderWidth: 4,
        ...px([t.spacing.sm, t.spacing.lg, t.spacing.xl]),
        paddingBottom: t.spacing.lg,
        paddingTop: [t.spacing.sm, t.spacing.md, t.spacing.lg],
        textAlign: "center",
      }),
      _css
    )}
  >
    <span
      css={s((t) =>
        ornament
          ? {}
          : {
              "&:before": {
                ...headingAlpha(t),
                color: t.color.background.feature5,
                content: "'“'",
                display: "block",
                fontSize: [80, null, 96],
                lineHeight: 0.75,
                marginBottom: "-0.25em",
              },
            }
      )}
      role="presentation"
    >
      {ornament}
    </span>
    <blockquote
      css={s(belt, headingCharlie, {
        "&:after": { content: "'”'" },
        "&:before": { content: "'“'" },
        fontSize: [20, null, 28],
        maxWidth: 640,
      })}
    >
      {children}
    </blockquote>
    {attribution && (
      <figcaption
        css={s(bodyText, (t) => ({
          "& > strong": { fontWeight: t.font.primary.weight.medium },
          marginTop: t.spacing.md,
        }))}
      >
        &ndash; {attribution}
      </figcaption>
    )}
  </figure>
);

export default Blockquote;
