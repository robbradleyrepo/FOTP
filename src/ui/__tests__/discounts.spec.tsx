import { captureException } from "@sentry/nextjs";
import { CART_STATE_VERSION } from "@sss/ecommerce/cart";
import { ParsedUrlQuery } from "querystring";
import React from "react";
import store from "store/dist/store.modern";
import { mocked } from "ts-jest/utils";

import { anonymous } from "../../sandbox";
import { ToastRack } from "../base/toast";

jest.mock("@sentry/nextjs", () => ({
  captureException: jest.fn(),
}));

// `DiscountListener` is mounted as part of the sandbox, so we'll mock the
// module everywhere but our `anonymous.component` call
jest.mock("../discounts", () => ({
  DiscountListener: () => null,
}));

const mockedCaptureException = mocked(captureException);

const mount = (params: ParsedUrlQuery = {}) => {
  const { DiscountListener } = jest.requireActual("../discounts");

  return anonymous.component({
    Component: (
      <>
        <DiscountListener />
        <ToastRack />
      </>
    ),
    params,
  });
};

describe("<DiscountListener />", () => {
  const DEFAULT_DISCOUNT_CODE = process.env.DEFAULT_DISCOUNT_CODE;
  const DISABLED_DISCOUNT_CODES = process.env.DISABLED_DISCOUNT_CODES;

  beforeEach(() => {
    process.env.DEFAULT_DISCOUNT_CODE = "DEFAULT";
  });

  afterEach(() => {
    process.env.DISABLED_DISCOUNT_CODES = DISABLED_DISCOUNT_CODES;
    jest.clearAllMocks();
  });

  afterAll(() => {
    process.env.DEFAULT_DISCOUNT_CODE = DEFAULT_DISCOUNT_CODE;
  });

  it("should automatically apply the default discount code", async () => {
    const { component } = await mount();

    expect(store.get("cart")).toMatchObject({ discountCode: "DEFAULT" });
    expect(
      component.getByText("DEFAULT has been automatically applied to your cart")
    ).toBeInTheDocument();
  });

  it("should automatically apply the discount code provided with the `aic` query param", async () => {
    const { component } = await mount({ aic: "TEST1" });

    expect(store.get("cart")).toMatchObject({ discountCode: "FRIEND-TEST1" });
    expect(
      component.getByText(
        "FRIEND-TEST1 has been automatically applied to your cart"
      )
    ).toBeInTheDocument();
  });

  it("should automatically apply the discount code provided with the `d` query param", async () => {
    const { component } = await mount({ d: "TEST2" });

    expect(store.get("cart")).toMatchObject({ discountCode: "TEST2" });
    expect(
      component.getByText("TEST2 has been automatically applied to your cart")
    ).toBeInTheDocument();
  });

  it("should automatically apply the discount code provided with the `discount` query param", async () => {
    const { component } = await mount({ discount: "TEST3" });

    expect(store.get("cart")).toMatchObject({ discountCode: "TEST3" });
    expect(
      component.getByText("TEST3 has been automatically applied to your cart")
    ).toBeInTheDocument();
  });

  it("should automatically remove any disabled codes", async () => {
    const disabledCodes = ["TEST1", "TEST2", "TEST3"];
    process.env.DISABLED_DISCOUNT_CODES = disabledCodes.join(",");

    for (const discountCode of disabledCodes) {
      store.set("cart", {
        checkoutId: null,
        customAttributes: {},
        discountCode,
        id: "da47f20d-9da2-472d-b922-e8e4eac54468",
        lineItems: [],
        rCheckoutId: null,
        version: CART_STATE_VERSION,
      });

      await mount();

      expect(store.get("cart")).toMatchObject({
        discountCode: process.env.DEFAULT_DISCOUNT_CODE,
      });
    }
  });

  it("should not automatically apply the default discount code if it is disabled", async () => {
    process.env.DISABLED_DISCOUNT_CODES = "DEFAULT";

    const { component } = await mount();

    expect(store.get("cart")).toMatchObject({
      discountCode: null,
    });
    expect(
      component.queryByText(
        "DEFAULT has been automatically applied to your cart"
      )
    ).toBeNull();
    expect(mockedCaptureException.mock.calls).toMatchSnapshot();
  });

  it("should not automatically apply the provided discount code if it is disabled", async () => {
    process.env.DISABLED_DISCOUNT_CODES = "FRIEND-TEST1, TEST2, TEST3";

    const testOne = await mount({ aic: "TEST1" });

    expect(store.get("cart")).toMatchObject({
      discountCode: process.env.DEFAULT_DISCOUNT_CODE,
    });
    expect(
      testOne.component.queryByText(
        "FRIEND-TEST1 has been automatically applied to your cart"
      )
    ).toBeNull();

    const testTwo = await mount({ d: "TEST2" });

    expect(store.get("cart")).toMatchObject({
      discountCode: process.env.DEFAULT_DISCOUNT_CODE,
    });
    expect(
      testTwo.component.queryByText(
        "TEST2 has been automatically applied to your cart"
      )
    ).toBeNull();

    const testThree = await mount({ discount: "TEST3" });

    expect(store.get("cart")).toMatchObject({
      discountCode: process.env.DEFAULT_DISCOUNT_CODE,
    });
    expect(
      testThree.component.queryByText(
        "TEST3 has been automatically applied to your cart"
      )
    ).toBeNull();
  });

  it("should apply the first active discount code if multiple query params are provided", async () => {
    const params = {
      aic: "TEST1",
      d: "TEST2",
      discount: "TEST3",
    };

    const testOne = await mount(params);

    expect(store.get("cart")).toMatchObject({ discountCode: "FRIEND-TEST1" });
    expect(
      testOne.component.getByText(
        "FRIEND-TEST1 has been automatically applied to your cart"
      )
    ).toBeInTheDocument();

    process.env.DISABLED_DISCOUNT_CODES = "FRIEND-TEST1";

    const testTwo = await mount(params);

    expect(store.get("cart")).toMatchObject({ discountCode: "TEST2" });
    expect(
      testTwo.component.getByText(
        "TEST2 has been automatically applied to your cart"
      )
    ).toBeInTheDocument();
  });

  it("should only notify the user if the applied code is different that in the cart", async () => {
    const cart = store.get("cart");
    store.set("cart", { ...cart, discountCode: "TEST" });

    const { component } = await mount({ discount: "TEST" });

    expect(store.get("cart")).toMatchObject({ discountCode: "TEST" });
    expect(
      component.queryByText("TEST has been automatically applied to your cart")
    ).toBeNull();
  });
});
