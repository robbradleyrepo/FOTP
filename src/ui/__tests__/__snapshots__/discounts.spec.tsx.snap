// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DiscountListener /> should not automatically apply the default discount code if it is disabled 1`] = `
Array [
  Array [
    "\\"DEFAULT_DISCOUNT_CODE\\" contains disabled discount code \\"DEFAULT\\"",
  ],
]
`;
